tiler/app.py

```diff
@@ -0,0 +1,404 @@
+from starlette.requests import Request
+from starlette.templating import Jinja2Templates
+from starlette_cramjam.middleware import CompressionMiddleware
+
+from titiler.application import __version__ as titiler_version
+from titiler.application.settings import ApiSettings
+from titiler.core.errors import DEFAULT_STATUS_CODES, add_exception_handlers
+from titiler.core.factory import (
+    AlgorithmFactory,
+    ColorMapFactory,
+    TilerFactory,
+    TMSFactory,
+)
+from titiler.core.middleware import (
+    CacheControlMiddleware,
+    LoggerMiddleware,
+    TotalTimeMiddleware,
+)
+from titiler.core.models.OGC import Conformance, Landing
+from titiler.core.resources.enums import MediaType
+from titiler.core.utils import accept_media_type, create_html_response, update_openapi
+from pydantic_settings import BaseSettings, SettingsConfigDict
+
+api_settings = ApiSettings()
+
+templates_location = [
+    jinja2.PackageLoader("titiler.application", "templates"),
+    jinja2.PackageLoader("titiler.core", "templates"),
+]
+
+jinja2_env = jinja2.Environment(
+    autoescape=jinja2.select_autoescape(["html", "xml"]),
+    loader=jinja2.ChoiceLoader(templates_location),
+)
+titiler_templates = Jinja2Templates(env=jinja2_env)
+
+
+###############################################################################
+
+app = FastAPI(
+    title=api_settings.name,
+    openapi_url="/api",
+    docs_url="/api.html",
+    description=api_settings.description,
+    version=titiler_version,
+    root_path=api_settings.root_path,
+)
+
+# Fix OpenAPI response header for OGC Common compatibility
+update_openapi(app)
+
+TITILER_CONFORMS_TO = {
+    "http://www.opengis.net/spec/ogcapi-common-1/1.0/conf/core",
+    "http://www.opengis.net/spec/ogcapi-common-1/1.0/conf/landing-page",
+    "http://www.opengis.net/spec/ogcapi-common-1/1.0/conf/oas30",
+    "http://www.opengis.net/spec/ogcapi-common-1/1.0/conf/html",
+    "http://www.opengis.net/spec/ogcapi-common-1/1.0/conf/json",
+}
+
+
+###############################################################################
+# TiTiler endpoints
+def DatasetPathParams(
+    dataset_id: Annotated[
+        Literal["coral", "other"],
+        Path(description="Dataset"),
+    ],
+    year: Annotated[int, Path(description="Year")],
+) -> str:
+    """Custom Dataset Path Parameter which define dataset_id and year PATH parameter and return a VRT url."""
+    name: str
+    if dataset_id == "coral":
+        name = "nbgi_clipped"
+    elif dataset_id == "other":
+        raise HTTPException(status_code=404, detail="Not Found.")
+
+    return f"https://syd1.digitaloceanspaces.com/mis-geotiff-storage/production/raster/{name}_{year}.vrt"
+
+
+tiler = TilerFactory(
+    reader=Reader,
+    path_dependency=DatasetPathParams,
+    add_ogc_maps=True,
+    templates=titiler_templates,
+    router_prefix="/dataset/{dataset_id}/years/{year}",
+)
+app.include_router(tiler.router, prefix="/dataset/{dataset_id}/years/{year}")
```

Comment:
small titiler hack, we use a special notation for the endpoint prefix `/dataset/{dataset_id}/years/{year}`. This will tell FastAPI to look for Path parameters (define in our custom `DatasetPathParams`)